### PR TITLE
8302108: Clean up placeholder supername code

### PR DESCRIPTION
**Pull request by coleenp**
Please review change to make PlaceholderEntry::_supername a SymbolHandle to correctly refcount the symbol rather than have ad-hoc code.  It also moves the private functions to private and asserts invariant that find_and_remove must always find a matching entry.  Finally, this also more eagerly removes the supername symbol from the placeholder entry, so I have to fix the test in https://github.com/openjdk/jdk/pull/12491 before pushing this.
Tested with tier1-4.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302108](https://bugs.openjdk.org/browse/JDK-8302108): Clean up placeholder supername code

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [1d5fd3c0](https://git.openjdk.org/jdk/pull/12495/files/1d5fd3c0d2666c6cd71b8ff123a49d83b98a33f4)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12495/head:pull/12495` \
`$ git checkout pull/12495`

Update a local copy of the PR: \
`$ git checkout pull/12495` \
`$ git pull https://git.openjdk.org/jdk pull/12495/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12495`

View PR using the GUI difftool: \
`$ git pr show -t 12495`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12495.diff">https://git.openjdk.org/jdk/pull/12495.diff</a>

</details>

### CODE REVIEW

**Review by dholmes-ora — APPROVED** [review]
Took me a while to get my head around the SymbolHandle mechanics for ref counting, but once I did this looks good. :) Thanks.

**Comment by coleenp** [comment]
Thanks David.  I had printf code to verify what SymbolHandle refcounting did to understand it but it's better that it just works rather than ad-hoc ref counting.
Thanks for reviewing!

**Review by iklam — APPROVED** [review]
LGTM. Just a small nit.

File: src/hotspot/share/classfile/placeholders.hpp [hunk 1]
```diff
@@ -99,39 +99,28 @@ class PlaceholderEntry {
   void add_seen_thread(JavaThread* thread, PlaceholderTable::classloadAction action);
   bool remove_seen_thread(JavaThread* thread, PlaceholderTable::classloadAction action);
 
+  SeenThread*        superThreadQ()        const { return _superThreadQ; }
```
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/classfile/placeholders.hpp:102]: I really wanted them moved to the section of the class where the other private methods are.  This class isn't so overgrown that we need to keep switching from private to public, and I want the public methods to be together.
Java is nice in that you can have 'private' and 'public' keywords on each function, but I'd still like to keep the public ones together.

**Comment by coleenp** [comment]
Thanks for reviewing Ioi!
/integrate

**Pull request closed without merging**